syntax = "proto3";

package woodstock;

message Empty {

}

message PoolUnused {
  bytes sha256 = 1;
  uint64 size = 3;
  uint64 compressedSize = 4;
}

message PoolRefCount {
  bytes sha256 = 1;
  uint32 refCount = 2;
  uint64 size = 3;
  uint64 compressedSize = 4;
}


message FileManifestStat {
  uint32 ownerId = 1;
  uint32 groupId = 2;
  uint64 size = 3;
  uint64 compressedSize = 8;
  int64 lastRead = 4;
  int64 lastModified = 5;
  int64 created = 6;
  uint32 mode = 7;

  uint64 dev = 9;
  uint64 rdev = 10;

  uint64 ino = 11;
  uint64 nlink = 12;
}

enum FileManifestAclQualifier {
    UNDEFINED = 0;
    USER_OBJ = 1;
    GROUP_OBJ = 2;
    OTHER = 3;
    USER_ID = 4;
    GROUP_ID = 5;
    MASK = 6;
}

message FileManifestAcl {
  FileManifestAclQualifier qualifier = 1;
  uint32 id = 2;
  uint32 perm = 3;
}

message FileManifestXAttr {
  bytes key = 1;
  bytes value = 2;
}

// File manifest used to store the content of a file
message FileManifest {
  bytes path = 1;
  FileManifestStat stats = 2;
  repeated FileManifestXAttr xattr = 5;
  repeated FileManifestAcl acl = 6;
  repeated bytes chunks = 3;
  bytes symlink = 7;
}

enum EntryType {
  // Add a new file
  ADD = 0;
  // Modify a file
  MODIFY = 1;
  // Remove a file
  REMOVE = 2;
}

// Journal entry
message FileManifestJournalEntry {
  EntryType type = 1;
  FileManifest manifest = 2;
}

enum LogLevel {
  log = 0;
  error = 1;
  warn = 2;
  debug = 3;
  verbose = 4;
}

// Log entry
message LogEntry {
  LogLevel level = 1;
  string context = 2;
  string line = 3;
}

message RefreshCacheHeader {
  string sharePath = 1;
}

message RefreshCacheRequest {
  oneof field {
    RefreshCacheHeader header = 1;
    FileManifest fileManifest = 2;
  }
}

message LaunchBackupRequest {
  Share share = 1;
}

message ChunkInformation {
  bytes filename = 1;
  uint64 position = 2;
  uint32 size = 3;
}

message FileChunk {
  bytes data = 1;
}

message StreamLogRequest {
}

message ExecuteCommandRequest {
  string command = 1;
}

message ExecuteCommandReply {
  int32 code = 1;
  string stdout = 2;
  string stderr = 3;
}

message Share {
  string sharePath = 1;
  repeated string includes = 2;
  repeated string excludes = 3;
}

message AuthenticateRequest {
  uint32 version = 1;
  string token = 2;
}

message AuthenticateReply {
  string sessionId = 3;
}

service WoodstockClientService {
  // 0. Authenticate
  rpc Authenticate(AuthenticateRequest) returns (AuthenticateReply);

  // 1. Execute a command directly on the client
  rpc ExecuteCommand(ExecuteCommandRequest) returns (ExecuteCommandReply) {}

  // 2. The server send the last cache version to the client
  rpc RefreshCache(stream RefreshCacheRequest) returns (Empty) {}

  // 3. The server launch the backup with a new number. The client will browse the client computer
  //    and will send all new file to the server
  rpc LaunchBackup(LaunchBackupRequest) returns (stream FileManifestJournalEntry) {}

  // 3. When the server receive a journal entry, the server will compare the chunk with the manifest
  //    and ask all necessary chunk to the client.
  rpc GetChunk(ChunkInformation) returns (stream FileChunk) {}

  // In parallele, the server will open a stream for the client to send log file.
  rpc StreamLog(StreamLogRequest) returns (stream LogEntry) {}

  // Free the client
  rpc CloseBackup(Empty) returns (Empty) {}
}
